File: LEETCODE/LEETCODE-CPP/695.cpp
//695. Max Area of Island
//https://leetcode.com/problems/max-area-of-island/description/

class Solution {
public:
    int dx[4]={-1,1,0,0};
    int dy[4]={0,0,-1,1};

    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int n1=grid.size(), n2=grid[0].size(), maxArea=0;

        vector<vector<bool>>visited(n1, vector<bool>(n2,false));

        for(int i=0;i<n1;i++){
            for(int j=0;j<n2;j++){
                if(grid[i][j]==1 && !visited[i][j]){
                    maxArea=max(maxArea, helper(grid,n1,n2,i,j,visited));
                }
            }
        }

        return maxArea;
    }

    int helper(vector<vector<int>>& grid, int n1, int n2, int i, int j, vector<vector<bool>>& visited){
        if(i<0 || j<0 || i>=n1 || j>=n2 || grid[i][j]==0 || visited[i][j])
            return 0;

        visited[i][j]=true;
        int area=1;
        for(int k=0;k<4;k++){
            area+=helper(grid,n1,n2,dx[k]+i,dy[k]+j,visited);
        }

        return area;
    }
};

File: LEETCODE/LEETCODE-PYTHON/695.py
# 695. Max Area of Island
# https://leetcode.com/problems/max-area-of-island/description/

# Method-1 using dfs
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m,n=len(grid), len(grid[0])
        max_area=0
        dir=[(0,1),(0,-1),(1,0),(-1,0)]

        def dfs(i,j):
            if i<0 or i>=m or j<0 or j>=n or grid[i][j]==0:
                return 0

            area=1
            grid[i][j]=0

            for dx,dy in dir:
                area+=dfs(i+dx,j+dy)

            return area

        for i in range(m):
            for j in range(n):
                if grid[i][j]==1:
                    area=dfs(i,j)

                    max_area=max(max_area,area)

        return max_area
    
#Method-2 using bfs
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m,n=len(grid), len(grid[0])
        max_area=0
        dir=[(0,1),(0,-1),(1,0),(-1,0)]

        def bfs(i,j):
            q=deque()
            q.append((i,j))

            area=1
            grid[i][j]=0

            while q:
                x,y=q.popleft()

                for dx, dy in dir:
                    temp_x, temp_y = x+dx, y+dy

                    if 0<=temp_x<m and 0<=temp_y<n and grid[temp_x][temp_y]==1:
                        area+=1
                        grid[temp_x][temp_y]=0
                        q.append((temp_x, temp_y))

            return area    

        for i in range(m):
            for j in range(n):
                if grid[i][j]==1:
                    area=bfs(i,j)

                    max_area=max(max_area,area)

        return max_area